Tree-comparison cleanup, rule convergence, and bug grouping

JJponce0913 opened the pull request:
## Tree-comparison cleanup, rule convergence, and bug grouping

**Bug grouping and convergence (new)**

- Added `src/tooling/sortBug.py` to support bug-family grouping and iterative convergence:
  - Scans known bug groups and loads their merged trees (`merged_tree.pkl`).
  - Tests whether two groups can be merged by generating a candidate rule from merged start nodes.
  - Validates candidate rules against:
    - both source known-bug groups (must fully match),
    - safe dataset (must not match).
  - Merges accepted groups on disk (`merge_groups_on_disk(...)`) and recomputes merged artifacts (`recompute_group_merged_tree(...)`).
  - Provides convergence helpers (`find_group_merges(...)`, `merge_converge_known_bugs(...)`) for repeated merge passes.
- Rule usability gating prevents weak rules from being accepted:
  - requires non-trivial pattern size,
  - requires meaningful modified-style signal (not only `"diff"` values).

---

**New tree/rule tooling scripts**

- Added `src/tooling/tree.py` to:
  - Load bug PKLs and convert `run_subject` data into comparable tree structures.
  - Extract tag/style signatures from merged trees.
  - Generate rules from merged trees (`create_rule(...)`).
  - Evaluate rules across folders (`check_all_pkls(...)`) for known vs safe sets.
- Added `src/tooling/treeComparison.py` to:
  - Define `Node` / `TextNode` models for comparison.
  - Convert `run_subject` -> node trees (`run_subject_to_node_tree(...)`).
  - Merge trees with shared-style intersection semantics.
  - Print/debug merged structures (`walk_tree(...)`, `walk_tree_verbose(...)`).

---

**Runner + skip-rule integration**

- Updated `src/lqc_selenium/runner.py` to:
  - Integrate rule-based skip checks through `should_skip(...)` from tree tooling.
  - Preserve the pre-minify subject and return it through minify flow.
  - Improve bug-path logging (`pattern found` vs `no pattern found`).
- Minify flow now:
  - Runs skip detection before full reduction.
  - Returns `(minified_run_subject, minified_run_result, prerun_subject, shouldSkip)` for downstream reporting.

---

**Bug-report saving changes**

- Updated `src/lqc_selenium/report/bug_report_helper.py` to:
  - Save report artifacts under a normalized bug-report folder.
  - Persist both:
    - `minified_run_subject.pkl`
    - `run_subject_prerun.pkl`
  - Include both minified and prerun subjects in JSON payload metadata.
  - Preserve `shouldSkip` state in report data.

---

**Minify replay/debug tooling**

- Added `src/tooling/run_minify.py` to:
  - Load a saved `run_subject.pkl`.
  - Replay Selenium + minification flow outside full fuzz generation.
  - Save debug artifacts and a summary under `bug_reports/debug_minify_runs/run-<id>`.
  - Optionally generate a bug report after variant testing.

---

**Tooling cleanup and script reorganization**

- Removed legacy tooling files:
  - `src/tooling/minifyTest.py`
  - `src/tooling/rules.py`
- Added helper scripts under `src/tooling/scripts/`:
  - `merge_folder.py`
  - `merge_folder_comparer.py`
  - `pickle_to_html.py`
  - `pickle_to_tree.py`

JJponce0913 committed:
add

JJponce0913 committed:
print JS diff

JJponce0913 committed:
add pickles

JJponce0913 committed:
Merge branch 'master' of https://github.com/JJponce0913/layout-quickcheck

JJponce0913 committed:
Enhance minification debugging and bug report workflow

Introduces a Manipulation class for clearer minimization steps and improves the minification process with step-by-step HTML output for debugging. Updates bug report saving to include pickle addresses, adds more detailed logging and error handling in layout testing, and refines the minification and test runner logic for better traceability and reproducibility. Also updates .gitignore to exclude new debug and test output files.

JJponce0913 committed:
Add bug report cases and debugging tools

Added transfer folder. Fixed the minification. Updated .gitignore to exclude /temp. Removed obsolete files and made minor updates to test and model scripts.

JJponce0913 committed:
Add Rule detection

JJponce0913 committed:
Update .gitignore

JJponce0913 committed:
Update Check_styles

JJponce0913 committed:
Add rule management and improve bug report handling

Introduces rule management via rules.json and src/lqc_selenium/rules.py, enabling rule addition and loading. Updates bug report logic to support custom timestamped folders when skipping, and refactors minification to check for specific patterns and style assignments. Minor code cleanups and disables verbose debug output in several modules.

JJponce0913 committed:
Remove obsolete bug report files and update tests

Deleted multiple old bug report files and style count aggregates from the Transfer directory. Migrated select bug report cases to shouldSkip-bug-report paths and added new test HTML files. Updated src/lqc_selenium/rules.py and runner.py to reflect these changes.

JJponce0913 committed:
copy over

JJponce0913 committed:
Postskip stuff

JJponce0913 committed:
Refactor rules and runner logic; update variants and config

Updated rules.json to redefine and expand display rules. Refactored runner.py to load and apply rules dynamically from JSON, improving rule matching logic.  Cleaned up and removed unused files. Minor config and bug report helper adjustments.

JJponce0913 committed:
Cleanup for PR Merge

JJponce0913 committed:
PR touchups

JJponce0913 committed:
Added comments to some tooling and finishing pr touchups.

pavpanchekha commented on src/lqc/config/file_config.py:
```diff
@@ -23,6 +23,8 @@ def __init__(self):
             os.makedirs(self.bug_report_file_dir)
     
+    def get(self, key, default=None):
+        return getattr(self, "data", {}).get(key, default)
```
What? Who calls this, and why does it exist? The code itself is kinda mysterious, it's basically doing `self.data.get(...)` but doesn't want to initialize `data`??

pavpanchekha commented on src/lqc/minify/minify_test_file.py:
```diff
@@ -11,14 +23,16 @@ def elements(tree):
             return proposed_run_subject
 
-        yield removeElement
+        yield Manipulation(removeElement, f"Remove element {element_id}")
```
What's the point of this? Are we going to parse it to get the element ID out at some point? Why not just do `removeElement.element_id = element_id` before returning it?

pavpanchekha commented on src/lqc/minify/minify_test_file.py:
```diff
@@ -239,18 +253,30 @@ def __init__(self):
+
+                #PRINT MANIPULATION
+                #print(manipulation)  # shows <Manipulation: Remove element X>
+                #print(f"Applying minimization step from {self.GENERATORS[self.current_generator_index].__name__}")
```
debug junk

pavpanchekha commented on src/lqc/model/run_result.py:
```diff
@@ -19,13 +19,22 @@ def isBug(self):
+    def isBug(self):
+        #BUG Consider Js Errors as non-bugs
+        return True """
+
```
Debug junk

pavpanchekha commented on src/lqc_selenium/selenium_harness/layout_tester.py:
```diff
@@ -22,29 +22,40 @@ def test_combination(webdriver, run_subject: RunSubject, slow=False, keep_file=F
+        results = webdriver.execute_script("return checkForBug()")
         
-        if slow: time.sleep(0.5)
+        #Printing the diff results
```
debug junk—maybe do a pass over the PR diff to find and remove these.

pavpanchekha commented on src/lqc_selenium/geckodriver.exe:
Please do not commit this

pavpanchekha commented on geckodriver.exe:
Please do not commit this either

pavpanchekha commented on .gitignore:
```diff
@@ -6,6 +6,7 @@ bugreportfiles
 bug_reports
 layoutfiles
 tmp_generated_files
+/temp
```
I don't think this does anything?

pavpanchekha commented on src/lqc_selenium/variants/variants.py:
```diff
@@ -121,27 +128,23 @@ def __repr__(self):
-            chrome_options.add_argument('--headless')
-            chrome_options.add_argument('--no-sandbox')
-            chrome_options.add_argument('--disable-dev-shm-usage')
+            chrome_options.add_argument("--headless=new")
```
This is, like, new syntax or something?

JJponce0913 replied:
Yeah, chrome driver wasn't working properly and also outputting debug info that didn't need to be printed.

pavpanchekha reviewed (changes requested):
I didn't finish reviewing but giving comments back now before I have to switch to another task. Maybe do a pass over the PR and remove the stuff that doesn't need to be committed.

JJponce0913 committed:
Add HTML tree comparison tooling and fix minify path

Introduces src/tooling/treeComparison.py for comparing HTML trees using BeautifulSoup, including node merging and tree walking utilities.

JJponce0913 committed:
Add __str__ methods and enhance tree comparison tooling

Implemented __str__ methods for ElementTree and StyleMap for improved readability. Updated treeComparison.py with enhanced debugging, sibling alignment, and more robust merging logic. Added checkPickle.py utility for inspecting pickled run_subject objects. Fixed default config path in runner.py.

JJponce0913 committed:
Revert "Add __str__ methods and enhance tree comparison tooling"

This reverts commit 8bf7a9f73777444500139bbe109c8fe2d6604f1a.

JJponce0913 committed:
Revert "Add HTML tree comparison tooling and fix minify path"

This reverts commit f4ead2002d1ef0c1a54eefaf29d202df3ebc0cc0.

JJponce0913 committed:
Remove unused code and update config,

Cleaned up .gitignore, deleted geckodriver.exe binaries, and various debugs.

JJponce0913 committed:
fix debug commenting

JJponce0913 committed:
Remove unused prints

Removed some more debug

JJponce0913 committed:
Change default config file for check_styles.py

JJponce0913 committed:
Update default config file

JJponce0913 committed:
Working Tree Comparison

JJponce0913 committed:
Merge remote-tracking branch 'origin/master' into TreeComparison

JJponce0913 committed:
Fixing the skip rules

JJponce0913 committed:
Fix Skipping

JJponce0913 committed:
Add rule convergence tooling and update pattern checks

Introduces ruleConvergence.py for merging and analyzing rule patterns Refactors check_pattern logic in runner.py for improved tag matching, updates treeComparison.py to fix dictionary merging and remove unused code.

JJponce0913 committed:
Update Convergence

JJponce0913 committed:
Fixed tooling imports

JJponce0913 committed:
Fix tooling

JJponce0913 committed:
Merge branch 'TreeComparison' of https://github.com/JJponce0913/layout-quickcheck into TreeComparison

JJponce0913 committed:
Fixed rule convergence check

JJponce0913 committed:
Fixing tooling.

JJponce0913 committed:
Refactor and clean up

Now succesfully use the same tree structure through runner.py. Refactored Rule convergence to match id's with both style and html patterns.

JJponce0913 committed:
Fix and started a sortBug

JJponce0913 committed:
Added basic known bug detection

JJponce0913 committed:
Fixing sortbug

JJponce0913 committed:
Fix sorting

JJponce0913 committed:
add base styles into rule creation

JJponce0913 committed:
get rid of uneeded tooling files

JJponce0913 committed:
document changes

JJponce0913 committed:
Documentation and refactors

JJponce0913 committed:
add docstring

JJponce0913 committed:
Merge branch 'master' into TreeComparisonCleanUp

JJponce0913 committed:
Merge pull request #1 from JJponce0913/TreeComparisonCleanUp

Tree comparison clean up to Master

pavpanchekha commented on src/lqc/config/file_config.py:
```diff
@@ -31,7 +31,15 @@ def getTimestampFilePath(self):
         return self.layout_file_dir, filepath, filename
     
-
+    def getCustomTimestampBugReport(self, custom_folder: str):
```
Any point to this? I think we're just renaming the folder; why not just modify `getTimestampFilePath` above?

JJponce0913 replied:
The getTimestampFilePath was specifically used just for tmp_files and creating a singuler timestamped .html and the getCustomTimestampBugReport was for creating a timestamped folder. I condensed it by adding an optional parameter of a specific filename to differentiate between wanting a timestamped file vs dir.

pavpanchekha commented on src/lqc_selenium/runner.py:
```diff
@@ -1,33 +1,142 @@
-    # Create final representations of minified files
     run_result, _ = test_combination(target_browser.getDriver(), run_subject)
-    return (run_subject, run_result)
+    return (run_subject, run_result,prerun_subject, shouldSkip)
```
```suggestion
    return (run_subject, run_result, prerun_subject, shouldSkip)
```

pavpanchekha commented on src/lqc_selenium/runner.py:
```diff
@@ -1,33 +1,142 @@
+    with redirect_stdout(io.StringIO()):
+        shouldSkip = should_skip(run_subject, rules)
+
+    #Skipe minimization if shouldSkip is True
```
```suggestion
    # Skip minimization if shouldSkip is True
```

pavpanchekha commented on src/lqc_selenium/runner.py:
```diff
@@ -1,33 +1,142 @@
+    conf = Config()
+    rules = conf.getRules()
+
+    with redirect_stdout(io.StringIO()):
```
Why do we need to redirect stdout? Seems like a code smell.

pavpanchekha replied:
Also if we removed the need for that we could inline `shouldSkip` which would be nicer.

JJponce0913 replied:
Leftover from debug stuff

pavpanchekha commented on src/lqc_selenium/runner.py:
```diff
@@ -1,33 +1,142 @@
 
 
 def minify(target_browser, run_subject):
+    prerun_subject = run_subject
```
Better to rename everything else to be, like `minimized_subject`

JJponce0913 replied:
Changed to minified_run_subject for consistensy with the rest of the code.

pavpanchekha commented on src/lqc_selenium/runner.py:
```diff
@@ -1,33 +1,142 @@
+        # If the minimized subject still triggers the bug, accept it as the new subject
         if run_result.isBug():
-            run_subject = proposed_run_subject
+            run_subject = temp_run_subject
```
Why rename this variable everywhere?

JJponce0913 replied:
not sure but changed it back

pavpanchekha commented on src/lqc_selenium/runner.py:
```diff
@@ -39,42 +148,38 @@ def find_bugs(counter):
-
         else:
             # Stage 2 - Minifying Bug
-            if run_result.type == BugType.PAGE_CRASH:
```
Why remove this? Was a useful informational message I think

pavpanchekha commented on src/lqc_selenium/runner.py:
All the new functions at the top of this file are dead code, why do we have them?

JJponce0913 replied:
Removed

pavpanchekha commented on src/lqc/rules/rule_engine.py:
I don't think this file should live in `tooling`, probably should go in `lqc` since it's backend-independent.

pavpanchekha replied:
Or maybe we just move should_skip? And everything else can stay in tooling?

JJponce0913 replied:
Moved to LQC under rules, so maybe future rule stuff can also live there if need, maybe move the merge tree stuff there as well?

pavpanchekha replied:
Yeah that makes sense to me, the merge tree stuff is again backend independent. Could be a follow-up PR but let's make sure to actually do the follow-up.

pavpanchekha commented on src/lqc/rules/tree_merge.py:
```diff
@@ -0,0 +1,330 @@
+import pickle
+
+class Node:
```
Why do we have a brand new tree type in here? I think LQC already has a tree type somewhere?

pavpanchekha replied:
I see, we want styles attached to nodes. Ideally all of LQC would switch to sticking the styles on nodes, it's a simpler & more useful format. But fine.

JJponce0913 replied:
So leave as a fix down the line?

pavpanchekha replied:
Agreed, but keep a list of these.

pavpanchekha commented on src/tooling/tree_merge.py:
```diff
@@ -0,0 +1,330 @@
+        if isinstance(v1, dict) and isinstance(v2, dict):
+            merged = _merge_dicts(v1, v2)
+            if merged:
+                out[k] = merged
```
I think if the merged dicts have no common keys this will accidentally drop it instead of saying "diff" or something.

JJponce0913 replied:
Yeah, was testing the effects of diff but added it back in.

pavpanchekha commented on src/tooling/tree_merge.py:
```diff
@@ -0,0 +1,330 @@
+        return TextNode(text, parent=parent)
+
+    if a_is_text or b_is_text:
+        return Node(tag="diff", id="empty", parent=parent, base_style={}, modified_style={})
```
Do we need the `id="empty"` here? Kind of ugly to have an extra sentinel object. Maybe just output None in this case?

JJponce0913 replied:
Yeah, i was worried that it might crash if set to none at some point but after review it shouldn't.

pavpanchekha commented on src/tooling/tree_merge.py:
```diff
@@ -0,0 +1,330 @@
+    tag = a.tag if a.tag == b.tag else "diff"
+    node_id = a.id if a.id == b.id else "diff"
+
+    base_style = _merge_dicts(getattr(a, "base_style", {}) or {}, getattr(b, "base_style", {}) or {})
```
Why do we have to use `getattr` here?

JJponce0913 replied:
was to ensure no crash while it wasn't fully implemented but fixed now.

pavpanchekha commented on src/lqc/rules/tree_merge.py:
```diff
@@ -0,0 +1,330 @@
+def pickle_to_node_tree(pickle_path):
+    with open(pickle_path, "rb") as f:
+        run_subject = pickle.load(f)
+    return run_subject_to_node_tree(run_subject)
```
I'm gonna mostly ignore this an the function it calls, they are long but I assume boring.

JJponce0913 replied:
correct

pavpanchekha commented on src/tooling/tree_merge.py:
```diff
@@ -0,0 +1,330 @@
+def walk_tree(n, depth=0):
+    indent = "  " * depth
+
+    if (isinstance(n, TextNode) or n.tag == "#text"):
```
We should dedup so that `#text` or `TextNode` is used but not both.

JJponce0913 replied:
Done

pavpanchekha commented on src/lqc/rules/tree_merge.py:
```diff
@@ -0,0 +1,330 @@
+        for child in n.children:
+            walk_tree(child, depth + 1)
+
+def walk_tree_verbose(n, depth=0):
```
Do we still need this>

JJponce0913 replied:
I think it could be useful down the line, but its not critical.

pavpanchekha reviewed (commented):
Mostly good but see comments, we should be able to merge after addressing. I didn't review the tooling scripts, I assume they're mostly for your use and some can go away after we have everything implemented and evaluated.

JJponce0913 committed:
Pr changes

Fixed text nodes, dead code and moved rule engine

JJponce0913 committed:
add back page cras

pavpanchekha commented on src/lqc_selenium/runner.py:
```diff
@@ -39,42 +63,41 @@ def find_bugs(counter):
-
-            (minified_run_subject, minified_run_result) = minify(target_browser, run_subject)
+                
+            (minified_run_subject, minified_run_result, prerun_subject, shouldSkip) = minify(target_browser, run_subject)
```
I don't think it makes any sense to return prerun_subject, it is the same exact object as run_subject

pavpanchekha commented on src/lqc/config/file_config.py:
```diff
@@ -22,18 +22,17 @@ def __init__(self):
+            os.makedirs(custom_dir)
+        dirpath = os.path.join(custom_dir, f"{custom_folder}-{formatted_timestamp}")
+        if filename:
+            filepath = os.path.join(dirpath, f"{formatted_timestamp}-{filename}")
```
Let the caller do this?

pavpanchekha commented on src/lqc/minify/minify_test_file.py:
```diff
@@ -11,16 +12,18 @@ def elements(tree):
 def Minify_RemoveEachElement(run_subject):
-
     for element in elements(run_subject.html_tree.tree):
+        element_id = element['id']
```
Got it but you don't need this line

pavpanchekha commented on src/lqc/minify/minify_test_file.py:
```diff
@@ -11,16 +12,18 @@ def elements(tree):
 
-        def removeElement(proposed_run_subject):
-            proposed_run_subject.removeElementById(element['id'])
+        def removeElement(proposed_run_subject, element_id=element_id):
```
```suggestion
        def removeElement(proposed_run_subject, element=element):
```

pavpanchekha commented on src/lqc/minify/minify_test_file.py:
```diff
@@ -11,16 +12,18 @@ def elements(tree):
-        def removeElement(proposed_run_subject):
-            proposed_run_subject.removeElementById(element['id'])
+        def removeElement(proposed_run_subject, element_id=element_id):
+            proposed_run_subject.removeElementById(element_id)
```
Then you don't need to change this line

pavpanchekha commented on src/lqc/rules/tree_merge.py:
Move this

JJponce0913 committed:
Formatting fixes

JJponce0913 committed:
more minor fixes for clarity
